File: binding/src/libqgraphicsitemanimation.cpp
#include <QGraphicsItem>
#include <QGraphicsItemAnimation>
#include <QObject>
#include <QPointF>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <QTimeLine>
#include <QTransform>
#include <qgraphicsitemanimation.h>
#include "libqgraphicsitemanimation.h"
#include "libqgraphicsitemanimation.hxx"

QGraphicsItemAnimation* QGraphicsItemAnimation_new() {
	 return new VirtualQGraphicsItemAnimation();
}

QGraphicsItemAnimation* QGraphicsItemAnimation_new2(QObject* parent) {
	 return new VirtualQGraphicsItemAnimation(parent);
}

libqt_string QGraphicsItemAnimation_Tr(const char* s) {
	QString _ret = QGraphicsItemAnimation::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

QGraphicsItem* QGraphicsItemAnimation_Item(const QGraphicsItemAnimation* self) {
	return self->item();
}

void QGraphicsItemAnimation_SetItem(QGraphicsItemAnimation* self, QGraphicsItem* item) {
	self->setItem(item);
}

QTimeLine* QGraphicsItemAnimation_TimeLine(const QGraphicsItemAnimation* self) {
	return self->timeLine();
}

void QGraphicsItemAnimation_SetTimeLine(QGraphicsItemAnimation* self, QTimeLine* timeLine) {
	self->setTimeLine(timeLine);
}

QPointF* QGraphicsItemAnimation_PosAt(const QGraphicsItemAnimation* self, double step) {
	return new QPointF(self->posAt(step));
}

libqt_list QGraphicsItemAnimation_PosList(const QGraphicsItemAnimation* self) {
	auto _ret = self->posList();
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		auto& _elem = _ret[_i];
		_data[_i] = new std::remove_reference_t<decltype(_elem)>(_elem);
	}
	return _arr;
}

void QGraphicsItemAnimation_SetPosAt(QGraphicsItemAnimation* self, double step, const QPointF* pos) {
	self->setPosAt(step, *pos);
}

QTransform* QGraphicsItemAnimation_TransformAt(const QGraphicsItemAnimation* self, double step) {
	return new QTransform(self->transformAt(step));
}

double QGraphicsItemAnimation_RotationAt(const QGraphicsItemAnimation* self, double step) {
	return self->rotationAt(step);
}

libqt_list QGraphicsItemAnimation_RotationList(const QGraphicsItemAnimation* self) {
	auto _ret = self->rotationList();
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		auto& _elem = _ret[_i];
		_data[_i] = new std::remove_reference_t<decltype(_elem)>(_elem);
	}
	return _arr;
}

void QGraphicsItemAnimation_SetRotationAt(QGraphicsItemAnimation* self, double step, double angle) {
	self->setRotationAt(step, angle);
}

double QGraphicsItemAnimation_XTranslationAt(const QGraphicsItemAnimation* self, double step) {
	return self->xTranslationAt(step);
}

double QGraphicsItemAnimation_YTranslationAt(const QGraphicsItemAnimation* self, double step) {
	return self->yTranslationAt(step);
}

libqt_list QGraphicsItemAnimation_TranslationList(const QGraphicsItemAnimation* self) {
	auto _ret = self->translationList();
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		auto& _elem = _ret[_i];
		_data[_i] = new std::remove_reference_t<decltype(_elem)>(_elem);
	}
	return _arr;
}

void QGraphicsItemAnimation_SetTranslationAt(QGraphicsItemAnimation* self, double step, double dx, double dy) {
	self->setTranslationAt(step, dx, dy);
}

double QGraphicsItemAnimation_VerticalScaleAt(const QGraphicsItemAnimation* self, double step) {
	return self->verticalScaleAt(step);
}

double QGraphicsItemAnimation_HorizontalScaleAt(const QGraphicsItemAnimation* self, double step) {
	return self->horizontalScaleAt(step);
}

libqt_list QGraphicsItemAnimation_ScaleList(const QGraphicsItemAnimation* self) {
	auto _ret = self->scaleList();
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		auto& _elem = _ret[_i];
		_data[_i] = new std::remove_reference_t<decltype(_elem)>(_elem);
	}
	return _arr;
}

void QGraphicsItemAnimation_SetScaleAt(QGraphicsItemAnimation* self, double step, double sx, double sy) {
	self->setScaleAt(step, sx, sy);
}

double QGraphicsItemAnimation_VerticalShearAt(const QGraphicsItemAnimation* self, double step) {
	return self->verticalShearAt(step);
}

double QGraphicsItemAnimation_HorizontalShearAt(const QGraphicsItemAnimation* self, double step) {
	return self->horizontalShearAt(step);
}

libqt_list QGraphicsItemAnimation_ShearList(const QGraphicsItemAnimation* self) {
	auto _ret = self->shearList();
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		auto& _elem = _ret[_i];
		_data[_i] = new std::remove_reference_t<decltype(_elem)>(_elem);
	}
	return _arr;
}

void QGraphicsItemAnimation_SetShearAt(QGraphicsItemAnimation* self, double step, double sh, double sv) {
	self->setShearAt(step, sh, sv);
}

void QGraphicsItemAnimation_Clear(QGraphicsItemAnimation* self) {
	self->clear();
}

void QGraphicsItemAnimation_SetStep(QGraphicsItemAnimation* self, double x) {
	self->setStep(x);
}

libqt_string QGraphicsItemAnimation_Tr2(const char* s, const char* c) {
	QString _ret = QGraphicsItemAnimation::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QGraphicsItemAnimation_Tr3(const char* s, const char* c, int n) {
	QString _ret = QGraphicsItemAnimation::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

// Derived class handler implementation
void QGraphicsItemAnimation_BeforeAnimationStep(QGraphicsItemAnimation* self, double step) {
	auto* vqgraphicsitemanimation = dynamic_cast<VirtualQGraphicsItemAnimation*>(self);
	if (vqgraphicsitemanimation && vqgraphicsitemanimation->isVirtualQGraphicsItemAnimation) {
	vqgraphicsitemanimation->beforeAnimationStep(step);
	} else {
	self->QGraphicsItemAnimation::beforeAnimationStep(step);
}
}

// Base class handler implementation
void QGraphicsItemAnimation_QBaseBeforeAnimationStep(QGraphicsItemAnimation* self, double step) {
	auto* vqgraphicsitemanimation = dynamic_cast<VirtualQGraphicsItemAnimation*>(self);
	if (vqgraphicsitemanimation && vqgraphicsitemanimation->isVirtualQGraphicsItemAnimation) {
vqgraphicsitemanimation->setQGraphicsItemAnimation_BeforeAnimationStep_IsBase(true);
	vqgraphicsitemanimation->beforeAnimationStep(step);
}
}

// Auxiliary method to allow providing re-implementation
void QGraphicsItemAnimation_OnBeforeAnimationStep(QGraphicsItemAnimation* self, intptr_t slot) {
	auto* vqgraphicsitemanimation = dynamic_cast<VirtualQGraphicsItemAnimation*>(self);
	if (vqgraphicsitemanimation && vqgraphicsitemanimation->isVirtualQGraphicsItemAnimation) {
vqgraphicsitemanimation->setQGraphicsItemAnimation_BeforeAnimationStep_Callback(reinterpret_cast<VirtualQGraphicsItemAnimation::QGraphicsItemAnimation_BeforeAnimationStep_Callback>(slot));
}
}

// Derived class handler implementation
void QGraphicsItemAnimation_AfterAnimationStep(QGraphicsItemAnimation* self, double step) {
	auto* vqgraphicsitemanimation = dynamic_cast<VirtualQGraphicsItemAnimation*>(self);
	if (vqgraphicsitemanimation && vqgraphicsitemanimation->isVirtualQGraphicsItemAnimation) {
	vqgraphicsitemanimation->afterAnimationStep(step);
	} else {
	self->QGraphicsItemAnimation::afterAnimationStep(step);
}
}

// Base class handler implementation
void QGraphicsItemAnimation_QBaseAfterAnimationStep(QGraphicsItemAnimation* self, double step) {
	auto* vqgraphicsitemanimation = dynamic_cast<VirtualQGraphicsItemAnimation*>(self);
	if (vqgraphicsitemanimation && vqgraphicsitemanimation->isVirtualQGraphicsItemAnimation) {
vqgraphicsitemanimation->setQGraphicsItemAnimation_AfterAnimationStep_IsBase(true);
	vqgraphicsitemanimation->afterAnimationStep(step);
}
}

// Auxiliary method to allow providing re-implementation
void QGraphicsItemAnimation_OnAfterAnimationStep(QGraphicsItemAnimation* self, intptr_t slot) {
	auto* vqgraphicsitemanimation = dynamic_cast<VirtualQGraphicsItemAnimation*>(self);
	if (vqgraphicsitemanimation && vqgraphicsitemanimation->isVirtualQGraphicsItemAnimation) {
vqgraphicsitemanimation->setQGraphicsItemAnimation_AfterAnimationStep_Callback(reinterpret_cast<VirtualQGraphicsItemAnimation::QGraphicsItemAnimation_AfterAnimationStep_Callback>(slot));
}
}

void QGraphicsItemAnimation_Delete(QGraphicsItemAnimation* self) {
    delete self;
}

File: binding/src/libqgraphicsitemanimation.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qgraphicsitemanimation.html
pub const qgraphicsitemanimation = struct {

    /// New constructs a new QGraphicsItemAnimation object.
    pub fn New() QtC.QGraphicsItemAnimation {
        return qtc.QGraphicsItemAnimation_new();
    }


    /// New2 constructs a new QGraphicsItemAnimation object.
    pub fn New2(parent: ?*anyopaque) QtC.QGraphicsItemAnimation {
        return qtc.QGraphicsItemAnimation_new2(@ptrCast(parent));
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QGraphicsItemAnimation_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qgraphicsitemanimation.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Item(self: ?*anyopaque, ) QtC.QGraphicsItem {
        return qtc.QGraphicsItemAnimation_Item(@ptrCast(self));
    }

    pub fn SetItem(self: ?*anyopaque, item: ?*anyopaque) void {
        qtc.QGraphicsItemAnimation_SetItem(@ptrCast(self), @ptrCast(item));
    }

    pub fn TimeLine(self: ?*anyopaque, ) QtC.QTimeLine {
        return qtc.QGraphicsItemAnimation_TimeLine(@ptrCast(self));
    }

    pub fn SetTimeLine(self: ?*anyopaque, timeLine: ?*anyopaque) void {
        qtc.QGraphicsItemAnimation_SetTimeLine(@ptrCast(self), @ptrCast(timeLine));
    }

    pub fn PosAt(self: ?*anyopaque, step: f64) QtC.QPointF {
        return qtc.QGraphicsItemAnimation_PosAt(@ptrCast(self), step);
    }

    pub fn PosList(self: ?*anyopaque, ) []const u8 {
        return qtc.QGraphicsItemAnimation_PosList(@ptrCast(self));
    }

    pub fn SetPosAt(self: ?*anyopaque, step: f64, pos: ?*anyopaque) void {
        qtc.QGraphicsItemAnimation_SetPosAt(@ptrCast(self), step, @ptrCast(pos));
    }

    pub fn TransformAt(self: ?*anyopaque, step: f64) QtC.QTransform {
        return qtc.QGraphicsItemAnimation_TransformAt(@ptrCast(self), step);
    }

    pub fn RotationAt(self: ?*anyopaque, step: f64) f64 {
        return qtc.QGraphicsItemAnimation_RotationAt(@ptrCast(self), step);
    }

    pub fn RotationList(self: ?*anyopaque, ) []const u8 {
        return qtc.QGraphicsItemAnimation_RotationList(@ptrCast(self));
    }

    pub fn SetRotationAt(self: ?*anyopaque, step: f64, angle: f64) void {
        qtc.QGraphicsItemAnimation_SetRotationAt(@ptrCast(self), step, angle);
    }

    pub fn XTranslationAt(self: ?*anyopaque, step: f64) f64 {
        return qtc.QGraphicsItemAnimation_XTranslationAt(@ptrCast(self), step);
    }

    pub fn YTranslationAt(self: ?*anyopaque, step: f64) f64 {
        return qtc.QGraphicsItemAnimation_YTranslationAt(@ptrCast(self), step);
    }

    pub fn TranslationList(self: ?*anyopaque, ) []const u8 {
        return qtc.QGraphicsItemAnimation_TranslationList(@ptrCast(self));
    }

    pub fn SetTranslationAt(self: ?*anyopaque, step: f64, dx: f64, dy: f64) void {
        qtc.QGraphicsItemAnimation_SetTranslationAt(@ptrCast(self), step, dx, dy);
    }

    pub fn VerticalScaleAt(self: ?*anyopaque, step: f64) f64 {
        return qtc.QGraphicsItemAnimation_VerticalScaleAt(@ptrCast(self), step);
    }

    pub fn HorizontalScaleAt(self: ?*anyopaque, step: f64) f64 {
        return qtc.QGraphicsItemAnimation_HorizontalScaleAt(@ptrCast(self), step);
    }

    pub fn ScaleList(self: ?*anyopaque, ) []const u8 {
        return qtc.QGraphicsItemAnimation_ScaleList(@ptrCast(self));
    }

    pub fn SetScaleAt(self: ?*anyopaque, step: f64, sx: f64, sy: f64) void {
        qtc.QGraphicsItemAnimation_SetScaleAt(@ptrCast(self), step, sx, sy);
    }

    pub fn VerticalShearAt(self: ?*anyopaque, step: f64) f64 {
        return qtc.QGraphicsItemAnimation_VerticalShearAt(@ptrCast(self), step);
    }

    pub fn HorizontalShearAt(self: ?*anyopaque, step: f64) f64 {
        return qtc.QGraphicsItemAnimation_HorizontalShearAt(@ptrCast(self), step);
    }

    pub fn ShearList(self: ?*anyopaque, ) []const u8 {
        return qtc.QGraphicsItemAnimation_ShearList(@ptrCast(self));
    }

    pub fn SetShearAt(self: ?*anyopaque, step: f64, sh: f64, sv: f64) void {
        qtc.QGraphicsItemAnimation_SetShearAt(@ptrCast(self), step, sh, sv);
    }

    pub fn Clear(self: ?*anyopaque, ) void {
        qtc.QGraphicsItemAnimation_Clear(@ptrCast(self));
    }

    pub fn SetStep(self: ?*anyopaque, x: f64) void {
        qtc.QGraphicsItemAnimation_SetStep(@ptrCast(self), x);
    }

    pub fn BeforeAnimationStep(self: ?*anyopaque, step: f64) void {
        qtc.QGraphicsItemAnimation_BeforeAnimationStep(@ptrCast(self), step);
    }

    /// Allows for overriding the related default method
    pub fn OnBeforeAnimationStep(self: ?*anyopaque, callback: *const fn (?*anyopaque, f64) callconv(.c) void) void {
        qtc.QGraphicsItemAnimation_OnBeforeAnimationStep(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseBeforeAnimationStep(self: ?*anyopaque, step: f64) void {
        qtc.QGraphicsItemAnimation_QBaseBeforeAnimationStep(@ptrCast(self), step);
    }

    pub fn AfterAnimationStep(self: ?*anyopaque, step: f64) void {
        qtc.QGraphicsItemAnimation_AfterAnimationStep(@ptrCast(self), step);
    }

    /// Allows for overriding the related default method
    pub fn OnAfterAnimationStep(self: ?*anyopaque, callback: *const fn (?*anyopaque, f64) callconv(.c) void) void {
        qtc.QGraphicsItemAnimation_OnAfterAnimationStep(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseAfterAnimationStep(self: ?*anyopaque, step: f64) void {
        qtc.QGraphicsItemAnimation_QBaseAfterAnimationStep(@ptrCast(self), step);
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QGraphicsItemAnimation_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qgraphicsitemanimation.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QGraphicsItemAnimation_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qgraphicsitemanimation.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QGraphicsItemAnimation_Delete(@ptrCast(self));
    }
};
